ddl/table.go

```diff
@@ -67,50 +67,36 @@ func onCreateTable(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error)
 		}
 		return ver, errors.Trace(err)
 	}
-	// Can not use both a placement policy and direct assignment. If you alter specify both in a CREATE TABLE or ALTER TABLE an error will be returned.
-	if tbInfo.DirectPlacementOpts != nil && tbInfo.PlacementPolicyRef != nil {
-		return ver, errors.Trace(ErrPlacementPolicyWithDirectOption.GenWithStackByArgs(tbInfo.PlacementPolicyRef.Name))
+	// placement rules meta inheritance.
+	dbInfo, err := checkSchemaExistAndCancelNotExistJob(t, job)
+	if err != nil {
+		return ver, errors.Trace(err)
 	}
-	var bundle *placement.Bundle
-	if tbInfo.DirectPlacementOpts != nil {
-		// check the direct placement option compatibility.
-		if err := checkPolicyValidation(tbInfo.DirectPlacementOpts); err != nil {
-			return ver, errors.Trace(err)
-		}
-		// build bundle from direct placement options.
-		bundle, err = placement.NewBundleFromOptions(tbInfo.DirectPlacementOpts)
-		if err != nil {
-			return ver, errors.Trace(err)
-		}
+	err = inheritPlacementRuleFromDB(tbInfo, dbInfo)
+	if err != nil {
+		return ver, errors.Trace(err)
 	}
-	if tbInfo.PlacementPolicyRef != nil {
-		// placement policy reference will override the direct placement options.
-		po, err := checkPlacementPolicyExistAndCancelNonExistJob(t, job, tbInfo.PlacementPolicyRef.ID)
-		if err != nil {
-			return ver, errors.Trace(infoschema.ErrPlacementPolicyNotExists.GenWithStackByArgs(tbInfo.PlacementPolicyRef.Name))
-		}
-		// build bundle from placement policy.
-		bundle, err = placement.NewBundleFromOptions(po.PlacementSettings)
-		if err != nil {
-			return ver, errors.Trace(err)
-		}
+
+	// build table & partition bundles if any.
+	tableBundle, err := getBundleFromTblInfo(t, job, tbInfo)
+	if err != nil {
+		return ver, errors.Trace(err)
 	}
-	if bundle == nil {
-		// get the default bundle from DB or PD.
-		bundle = infoschema.GetBundle(d.infoCache.GetLatest(), []int64{schemaID})
+	partitionBundles, err := getBundleFromPartitionDef(t, job, tbInfo)
+	if err != nil {
+		return ver, errors.Trace(err)
 	}
+	bundles := make([]*placement.Bundle, 0, 1+len(partitionBundles))
+	bundles = append(bundles, tableBundle)
+	bundles = append(bundles, partitionBundles...)
+
 	// Do the http request only when the rules is existed.
 	syncPlacementRules := func() error {
```

Comment:
I think this function can be inlined, since `Do the http request only when the rules is existed.` it is done by `infosync.PutRuleBundles`.